File: internal/store/redis_live.go
package store

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/clock"
	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/constants"
	"github.com/TrustEdgeOrg/TrustEdge-Agent-API/internal/models"
	"github.com/redis/go-redis/v9"
)

// DeviceLatest is the live per-device document TrustEdge reads.
type DeviceLatest struct {
	DeviceID       string         `json:"device_id"`
	LastSeenAt     *time.Time     `json:"last_seen_at,omitempty"`
	ClientDetails  map[string]any `json:"client_details,omitempty"`
	NetworkSummary map[string]any `json:"network_summary,omitempty"`
	ActionSummary  map[string]any `json:"action_summary,omitempty"`
}

type redisLive struct {
	client    *redis.Client
	clock     clock.Clock
	maxEvents int
	log       *log.Logger
}

func newRedisLive(redisURL string, maxEvents int, logger *log.Logger, clk clock.Clock) (*redisLive, error) {
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, fmt.Errorf("parse redis url: %w", err)
	}
	client := redis.NewClient(opts)
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	if err := client.Ping(ctx).Err(); err != nil {
		_ = client.Close()
		return nil, fmt.Errorf("redis ping: %w", err)
	}
	if logger == nil {
		logger = log.Default()
	}
	if clk == nil {
		clk = clock.Real{}
	}
	return &redisLive{client: client, clock: clk, maxEvents: maxEvents, log: logger}, nil
}

func (r *redisLive) Close() error {
	if r == nil || r.client == nil {
		return nil
	}
	return r.client.Close()
}

func latestKey(deviceID string) string { return fmt.Sprintf(constants.RedisLatestKeyFmt, deviceID) }
func eventsKey(deviceID string) string { return fmt.Sprintf(constants.RedisEventsKeyFmt, deviceID) }
func deviceAuthKey(deviceID string) string {
	return fmt.Sprintf("twin:device:%s:auth", deviceID)
}

func (r *redisLive) SaveDeviceAuth(rec *deviceRecord) error {
	if r == nil || rec == nil || rec.DeviceID == "" {
		return nil
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	data, err := json.Marshal(rec)
	if err != nil {
		return err
	}
	pipe := r.client.Pipeline()
	pipe.Set(ctx, deviceAuthKey(rec.DeviceID), data, 0)
	if rec.DeviceToken != "" {
		pipe.HSet(ctx, constants.RedisDeviceTokensKey, rec.DeviceToken, rec.DeviceID)
	}
	_, err = pipe.Exec(ctx)
	return err
}

func (r *redisLive) LoadDeviceAuth() ([]*deviceRecord, error) {
	if r == nil {
		return nil, nil
	}
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	tokenMap, err := r.client.HGetAll(ctx, constants.RedisDeviceTokensKey).Result()
	if err != nil {
		return nil, err
	}
	seen := map[string]struct{}{}
	var records []*deviceRecord
	for _, deviceID := range tokenMap {
		if deviceID == "" {
			continue
		}
		if _, ok := seen[deviceID]; ok {
			continue
		}
		seen[deviceID] = struct{}{}
		raw, err := r.client.Get(ctx, deviceAuthKey(deviceID)).Bytes()
		if err != nil || len(raw) == 0 {
			continue
		}
		var rec deviceRecord
		if err := json.Unmarshal(raw, &rec); err != nil {
			continue
		}
		records = append(records, &rec)
	}
	return records, nil
}

func (r *redisLive) loadLatest(ctx context.Context, deviceID string) DeviceLatest {
	raw, err := r.client.Get(ctx, latestKey(deviceID)).Bytes()
	if err != nil || len(raw) == 0 {
		return DeviceLatest{DeviceID: deviceID}
	}
	var doc DeviceLatest
	if err := json.Unmarshal(raw, &doc); err != nil {
		return DeviceLatest{DeviceID: deviceID}
	}
	doc.DeviceID = deviceID
	return doc
}

func (r *redisLive) saveLatest(ctx context.Context, doc DeviceLatest) error {
	data, err := json.Marshal(doc)
	if err != nil {
		return err
	}
	pipe := r.client.Pipeline()
	pipe.SAdd(ctx, constants.RedisDevicesKey, doc.DeviceID)
	pipe.Set(ctx, latestKey(doc.DeviceID), data, 0)
	_, err = pipe.Exec(ctx)
	return err
}

func (r *redisLive) UpsertRegister(deviceID string, details map[string]any, seenAt time.Time) {
	if r == nil {
		return
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	doc := r.loadLatest(ctx, deviceID)
	doc.LastSeenAt = &seenAt
	if doc.ClientDetails == nil {
		doc.ClientDetails = map[string]any{}
	}
	for k, v := range details {
		if v == nil {
			continue
		}
		if s, ok := v.(string); ok && s == "" {
			continue
		}
		doc.ClientDetails[k] = v
	}
	if err := r.saveLatest(ctx, doc); err != nil {
		r.log.Printf("redis register %s: %v", deviceID, err)
	}
}

func (r *redisLive) UpsertEvent(ev models.Event) {
	if r == nil {
		return
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()

	doc := r.loadLatest(ctx, ev.DeviceID)
	ts := ev.TS
	if ts.IsZero() {
		ts = r.clock.Now()
	}
	doc.LastSeenAt = &ts
	switch ev.Type {
	case constants.TypeClientDetails:
		doc.ClientDetails = ev.Payload
	case constants.TypeNetworkSummary:
		doc.NetworkSummary = ev.Payload
	case constants.TypeActionSummary:
		doc.ActionSummary = ev.Payload
	}

	data, err := json.Marshal(doc)
	if err != nil {
		r.log.Printf("redis event marshal %s: %v", ev.DeviceID, err)
		return
	}
	evData, err := json.Marshal(ev)
	if err != nil {
		r.log.Printf("redis event body %s: %v", ev.DeviceID, err)
		return
	}

	score := float64(ts.UnixMilli())
	ekey := eventsKey(ev.DeviceID)
	pipe := r.client.Pipeline()
	pipe.SAdd(ctx, constants.RedisDevicesKey, ev.DeviceID)
	pipe.Set(ctx, latestKey(ev.DeviceID), data, 0)
	pipe.ZAdd(ctx, ekey, redis.Z{Score: score, Member: string(evData)})
	if r.maxEvents > 0 {
		pipe.ZRemRangeByRank(ctx, ekey, 0, int64(-r.maxEvents-1))
	}
	if _, err := pipe.Exec(ctx); err != nil {
		r.log.Printf("redis event %s: %v", ev.DeviceID, err)
	}
}

File: app/redis_live.py
from __future__ import annotations

import json
import logging
from datetime import datetime
from typing import Any

import redis

from app import clock as clock_mod
from app.constants import (
    REDIS_DEVICE_TOKENS_KEY,
    REDIS_DEVICES_KEY,
    REDIS_EVENTS_KEY_FMT,
    REDIS_LATEST_KEY_FMT,
    TYPE_ACTION_SUMMARY,
    TYPE_CLIENT_DETAILS,
    TYPE_NETWORK_SUMMARY,
)
from app.models import Event

LOG = logging.getLogger("trustedge-agent-api")


class DeviceLatest:
    def __init__(
        self,
        device_id: str,
        last_seen_at: datetime | None = None,
        client_details: dict[str, Any] | None = None,
        network_summary: dict[str, Any] | None = None,
        action_summary: dict[str, Any] | None = None,
    ) -> None:
        self.device_id = device_id
        self.last_seen_at = last_seen_at
        self.client_details = client_details or {}
        self.network_summary = network_summary or {}
        self.action_summary = action_summary or {}

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"device_id": self.device_id}
        if self.last_seen_at is not None:
            out["last_seen_at"] = self.last_seen_at.isoformat().replace("+00:00", "Z")
        if self.client_details:
            out["client_details"] = self.client_details
        if self.network_summary:
            out["network_summary"] = self.network_summary
        if self.action_summary:
            out["action_summary"] = self.action_summary
        return out


def _latest_key(device_id: str) -> str:
    return REDIS_LATEST_KEY_FMT.format(device_id=device_id)


def _events_key(device_id: str) -> str:
    return REDIS_EVENTS_KEY_FMT.format(device_id=device_id)


def _device_auth_key(device_id: str) -> str:
    return f"twin:device:{device_id}:auth"


class RedisLive:
    def __init__(self, redis_url: str, max_events: int) -> None:
        self._client = redis.from_url(redis_url, decode_responses=False)
        self._client.ping()
        self._max_events = max_events

    def close(self) -> None:
        self._client.close()

    def save_device_auth(self, record: dict[str, Any]) -> None:
        device_id = str(record.get("device_id") or "")
        if not device_id:
            return
        pipe = self._client.pipeline()
        pipe.set(_device_auth_key(device_id), json.dumps(record))
        token = str(record.get("device_token") or "")
        if token:
            pipe.hset(REDIS_DEVICE_TOKENS_KEY, token, device_id)
        pipe.execute()

    def load_device_auth(self) -> list[dict[str, Any]]:
        token_map = self._client.hgetall(REDIS_DEVICE_TOKENS_KEY)
        seen: set[str] = set()
        records: list[dict[str, Any]] = []
        for device_id_raw in token_map.values():
            device_id = device_id_raw.decode() if isinstance(device_id_raw, bytes) else str(device_id_raw)
            if not device_id or device_id in seen:
                continue
            seen.add(device_id)
            raw = self._client.get(_device_auth_key(device_id))
            if not raw:
                continue
            try:
                data = json.loads(raw)
            except json.JSONDecodeError:
                continue
            if isinstance(data, dict):
                records.append(data)
        return records

    def _load_latest(self, device_id: str) -> DeviceLatest:
        raw = self._client.get(_latest_key(device_id))
        if not raw:
            return DeviceLatest(device_id=device_id)
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            return DeviceLatest(device_id=device_id)
        if not isinstance(data, dict):
            return DeviceLatest(device_id=device_id)
        last_seen = data.get("last_seen_at")
        parsed_seen: datetime | None = None
        if isinstance(last_seen, str) and last_seen:
            text = last_seen[:-1] + "+00:00" if last_seen.endswith("Z") else last_seen
            try:
                parsed_seen = datetime.fromisoformat(text)
            except ValueError:
                parsed_seen = None
        return DeviceLatest(
            device_id=device_id,
            last_seen_at=parsed_seen,
            client_details=data.get("client_details") if isinstance(data.get("client_details"), dict) else {},
            network_summary=data.get("network_summary") if isinstance(data.get("network_summary"), dict) else {},
            action_summary=data.get("action_summary") if isinstance(data.get("action_summary"), dict) else {},
        )

    def _save_latest(self, doc: DeviceLatest) -> None:
        pipe = self._client.pipeline()
        pipe.sadd(REDIS_DEVICES_KEY, doc.device_id)
        pipe.set(_latest_key(doc.device_id), json.dumps(doc.to_dict()))
        pipe.execute()

    def upsert_register(self, device_id: str, details: dict[str, Any], seen_at: datetime) -> None:
        doc = self._load_latest(device_id)
        doc.last_seen_at = seen_at
        for key, value in details.items():
            if value is None:
                continue
            if isinstance(value, str) and value == "":
                continue
            doc.client_details[key] = value
        try:
            self._save_latest(doc)
        except redis.RedisError as exc:
            LOG.warning("redis register %s: %s", device_id, exc)

    def upsert_event(self, event: Event) -> None:
        doc = self._load_latest(event.device_id)
        ts = event.ts or clock_mod.now_utc()
        doc.last_seen_at = ts
        if event.type == TYPE_CLIENT_DETAILS:
            doc.client_details = dict(event.payload)
        elif event.type == TYPE_NETWORK_SUMMARY:
            doc.network_summary = dict(event.payload)
        elif event.type == TYPE_ACTION_SUMMARY:
            doc.action_summary = dict(event.payload)

        ev_data = json.dumps(event.model_dump(mode="json", by_alias=False))
        score = ts.timestamp() * 1000.0
        pipe = self._client.pipeline()
        pipe.sadd(REDIS_DEVICES_KEY, event.device_id)
        pipe.set(_latest_key(event.device_id), json.dumps(doc.to_dict()))
        pipe.zadd(_events_key(event.device_id), {ev_data: score})
        if self._max_events > 0:
            pipe.zremrangebyrank(_events_key(event.device_id), 0, -(self._max_events + 1))
        try:
            pipe.execute()
        except redis.RedisError as exc:
            LOG.warning("redis event %s: %s", event.device_id, exc)
